src/shmem_atomic.h

```diff
@@ -25,6 +25,8 @@
 #endif
 
 #define COMPILER_FENCE() do { __asm__ __volatile__ ("" ::: "memory"); } while (0)
+#define LFENCE()  __asm__ __volatile__ ("lfence" ::: "memory")
```

Comment:
Need ```do { ... } while (0)``` around function-like macros to safely eat the trailing semicolon: https://gcc.gnu.org/onlinedocs/cpp/Swallowing-the-Semicolon.html

Reply:
If you structure it this way, the ```LFENCE``` and ```SFENCE``` macros should also be defined within an ```#if defined(__i386__) || defined(__x86_64__)``` block with an else case that defines them to ```__sync_synchronize```.  I think I would prefer to just put this directly into the membar_load/store routines. 
 Since lfence/sfence are only used once, there's limited value to defining macros for them.

Reply:
Thanks @jdinan. So, similar to this? In this case, we should not need #if defined(__i386__) || defined(__x86_64__).
shmem_internal_membar_load(void) {
    do { 
          __asm__ __volatile__ ("lfence" ::: "memory"); 
     } while (0);
    return;
 }

Reply:
The ```do { ... } while (0)``` business is for creating function-like macros, so not needed if you stick the asm into the membar_load function.  I think the logic might be simpler if you do:
```
shmem_internal_membar_load(void) {
#if defined(__i386__) || defined(__x86_64__)
    __asm__ __volatile__ ("lfence" ::: "memory"); 
#else
    __sync_synchronize();
#endif
    return;
 }
```

Reply:
Ok. Got it. Thanks.